File: application/api/forms/key-event.js
/* global Java */

define(['boxing'], function(B) {
    var className = "com.eas.client.forms.events.KeyEvent";
    var javaClass = Java.type(className);
    /**
     * Generated constructor.
     * @constructor KeyEvent KeyEvent
     */
    function KeyEvent() {
        var maxArgs = 0;
        var delegate = arguments.length > maxArgs ?
              arguments[maxArgs] 
            : new javaClass();

        Object.defineProperty(this, "unwrap", {
            configurable: true,
            value: function() {
                return delegate;
            }
        });
        if(KeyEvent.superclass)
            KeyEvent.superclass.constructor.apply(this, arguments);
        delegate.setPublished(this);
        /**
         * Alt key is down on this event.
         */
        this.altDown = true;
        Object.defineProperty(this, "altDown", {
            get: function() {
                var value = delegate.altDown;
                return B.boxAsJs(value);
            }
        });

        /**
         * Ctrl key is down on this event.
         */
        this.controlDown = true;
        Object.defineProperty(this, "controlDown", {
            get: function() {
                var value = delegate.controlDown;
                return B.boxAsJs(value);
            }
        });

        /**
         * Shift key is down on this event.
         */
        this.shiftDown = true;
        Object.defineProperty(this, "shiftDown", {
            get: function() {
                var value = delegate.shiftDown;
                return B.boxAsJs(value);
            }
        });

        /**
         * Char associated with this event.
         */
        this.char = '';
        Object.defineProperty(this, "char", {
            get: function() {
                var value = delegate.char;
                return B.boxAsJs(value);
            }
        });

        /**
         * Meta key is down on this event.
         */
        this.metaDown = true;
        Object.defineProperty(this, "metaDown", {
            get: function() {
                var value = delegate.metaDown;
                return B.boxAsJs(value);
            }
        });

        /**
         * The source object of the event.
         */
        this.source = new Object();
        Object.defineProperty(this, "source", {
            get: function() {
                var value = delegate.source;
                return B.boxAsJs(value);
            }
        });

        /**
         * Key code associated with this event.
         */
        this.key = 0;
        Object.defineProperty(this, "key", {
            get: function() {
                var value = delegate.key;
                return B.boxAsJs(value);
            }
        });

    }

    var ScriptsClass = Java.type("com.eas.script.Scripts");
    var space = ScriptsClass.getSpace();
    space.putPublisher(className, function(aDelegate) {
        return new KeyEvent(aDelegate);
    });
    return KeyEvent;
});

File: forms/src/com/eas/client/forms/events/KeyEvent.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package com.eas.client.forms.events;

import com.eas.script.NoPublisherException;
import com.eas.script.ScriptFunction;
import com.eas.script.Scripts;
import jdk.nashorn.api.scripting.JSObject;

/**
 *
 * @author mg
 */
public class KeyEvent extends Event<java.awt.event.KeyEvent> {

    protected KeyEvent(java.awt.event.KeyEvent aDelegate) {
        super(aDelegate);
    }

    private static final String KEY_JS_DOC = ""
            + "/**\n"
            + " * Key code associated with this event.\n"
            + " */";

    @ScriptFunction(jsDoc = KEY_JS_DOC)
    public int getKey() {
        return delegate.getKeyCode();
    }

    private static final String ALT_DOWN_JS_DOC = ""
            + "/**\n"
            + " * Alt key is down on this event.\n"
            + " */";

    @ScriptFunction(jsDoc = ALT_DOWN_JS_DOC)
    public boolean isAltDown() {
        return delegate.isAltDown() || delegate.isAltGraphDown();
    }

    private static final String CONTROL_DOWN_JS_DOC = ""
            + "/**\n"
            + " * Ctrl key is down on this event.\n"
            + " */";

    @ScriptFunction(jsDoc = CONTROL_DOWN_JS_DOC)
    public boolean isControlDown() {
        return delegate.isControlDown();
    }

    private static final String SHIFT_DOWN_JS_DOC = ""
            + "/**\n"
            + " * Shift key is down on this event.\n"
            + " */";

    @ScriptFunction(jsDoc = SHIFT_DOWN_JS_DOC)
    public boolean isShiftDown() {
        return delegate.isShiftDown();
    }

    private static final String META_DOWN_JS_DOC = ""
            + "/**\n"
            + " * Meta key is down on this event.\n"
            + " */";

    @ScriptFunction(jsDoc = META_DOWN_JS_DOC)
    public boolean isMetaDown() {
        return delegate.isMetaDown();
    }

    private static final String KEY_CHAR_JS_DOC = ""
            + "/**\n"
            + " * Char associated with this event.\n"
            + " */";

    @ScriptFunction(jsDoc = KEY_CHAR_JS_DOC)
    public String getChar() {
        char res = delegate.getKeyChar();
        return res == java.awt.event.KeyEvent.CHAR_UNDEFINED ? "" : Character.toString(res);
    }

    @Override
    public JSObject getPublished() {
        if (published == null) {
            JSObject publisher = Scripts.getSpace().getPublisher(this.getClass().getName());
            if (publisher == null || !publisher.isFunction()) {
                throw new NoPublisherException();
            }
            published = (JSObject) publisher.call(null, new Object[]{this});
        }
        return published;
    }
}
